packages/next/src/seo/route/index.ts

```diff
@@ -0,0 +1,155 @@
+import type { Metadata } from 'next';
+import type { OpenGraphMetadata } from './opengraph-metadata/types';
+import type { TwitterMetadata } from './twitter-metadata/types';
+import type { SiteMetadata } from './site-metadata/types';
+import type { RouteMetadataGeneratorPlugin } from './types';
+
+import fetchRouteOpenGraphMetadata from './opengraph-metadata/fetcher';
+import validateRouteOpenGraphMetadata from './opengraph-metadata/validator';
+import parseRouteOpenGraphMetadata from './opengraph-metadata/parser';
+
+import fetchRouteTwitterMetadata from './twitter-metadata/fetcher';
+import validateRouteTwitterMetadata from './twitter-metadata/validator';
+import parseRouteTwitterMetadata from './twitter-metadata/parser';
+import fetchRouteSiteMetadata from './site-metadata/fetcher';
+import parseRouteSiteMetadata from './site-metadata/parser';
+import validateRouteSiteMetadata from './site-metadata/validator';
+
+const plugins: Record< string, RouteMetadataGeneratorPlugin< unknown > > = {};
+
+/**
+ *
+ * @param generator
+ * @param key
+ */
+export function registerMetadataGenerator< T extends unknown >(
+	generator: RouteMetadataGeneratorPlugin< T >,
+	key: string
+): void {
+	plugins[ key ] = generator as RouteMetadataGeneratorPlugin< unknown >;
+}
+
+const OpenGraphMetaDataGenerator: RouteMetadataGeneratorPlugin< OpenGraphMetadata > =
+	{
+		fetcher: fetchRouteOpenGraphMetadata,
+		validator: validateRouteOpenGraphMetadata,
+		parser: parseRouteOpenGraphMetadata,
+	};
+
+const TwitterMetaDataGenerator: RouteMetadataGeneratorPlugin< TwitterMetadata > =
+	{
+		fetcher: fetchRouteTwitterMetadata,
+		validator: validateRouteTwitterMetadata,
+		parser: parseRouteTwitterMetadata,
+	};
+
+const SiteMetaDataGenerator: RouteMetadataGeneratorPlugin< SiteMetadata > = {
+	fetcher: fetchRouteSiteMetadata,
+	validator: validateRouteSiteMetadata,
+	parser: parseRouteSiteMetadata,
+};
+
+const DEFAULT_FETCHED_OBJECT = {};
+const DEFAULT_VALIDATED_OBJECT = {};
+const DEFAULT_PARSED_OBJECT = {};
+
+registerMetadataGenerator( OpenGraphMetaDataGenerator, 'open-graph' );
+registerMetadataGenerator( TwitterMetaDataGenerator, 'twitter' );
+registerMetadataGenerator( SiteMetaDataGenerator, 'site' );
```

Comment:
Maybe its the lack of docs on this file but I am not understanding the api, nor what\how the lifecycle happens or what functionality is pluggable.

Reply:
I figured out what I dont like:

These `register*` calls are side effects, but there's multiple exportable things in this file not related to those side effects.

We need to separate the concerns somehow.

Reply:
This will be refactored into a class. Register calls would be wrapped in a constructor. Plugins will be added through config management in the constructor.